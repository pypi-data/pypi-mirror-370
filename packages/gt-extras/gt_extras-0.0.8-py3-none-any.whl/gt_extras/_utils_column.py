from __future__ import annotations

import warnings

from great_tables import GT
from great_tables._data_color.base import _rescale_numeric
from great_tables._locations import resolve_cols_c
from great_tables._tbl_data import SelectExpr, is_na, to_list

__all__ = [
    "_validate_and_get_single_column",
    "_scale_numeric_column",
    "_format_numeric_text",
]


def _validate_and_get_single_column(
    gt: GT,
    expr: SelectExpr,
) -> tuple[str, list]:
    """
    Validate that expr resolves to a single column and return the column name and values.

    Parameters
    ----------
    gt
        The `GT` object containing the data
    expr
        The column expression to resolve

    Returns
    -------
    tuple[str, list]
        A tuple of (column_name, column_values)

    Raises
    ------
    KeyError
        If the column is not found
    ValueError
        If multiple columns are resolved
    """
    col_names = resolve_cols_c(data=gt, expr=expr)

    if len(col_names) == 0:
        raise KeyError(f"Column '{expr}' not found in the table.")
    if len(col_names) > 1:
        raise ValueError(
            f"Expected a single column, but got multiple columns: {col_names}"
        )

    col_name = col_names[0]
    col_vals = to_list(gt._tbl_data[col_name])

    return col_name, col_vals


def _scale_numeric_column(
    data_table,
    col_name: str,
    col_vals: list,
    domain: list[float] | list[int] | None = None,
    default_domain_min_zero: bool = True,
) -> list[float]:
    """
    Process and scale a numeric column, handling NA values.

    Parameters
    ----------
    data_table
        The underlying data table
    col_name
        Name of the column (for error messages)
    col_vals
        The column values
    domain
        The domain for scaling. If None, uses a default domain, based on `default_domain_min_zero`
    default_domain_min_zero
        If true, the defsult domain will be [0, max(values)], otherwise [min(values), max(values)]
    Returns
    -------
    list[float]
        Scaled values with NAs mapped to 0, values above the domains mapped to 1,
        and values below the domain mapped to 0.

    Raises
    ------
    TypeError
        If the column is not numeric
    """
    col_vals_filtered = [x for x in col_vals if not is_na(data_table, x)]

    # Check that column has numeric data
    if len(col_vals_filtered) and all(
        isinstance(x, (int, float)) for x in col_vals_filtered
    ):
        # If `domain` is not provided, then set it to a default domain
        if domain is None:
            if default_domain_min_zero:
                domain = [0, max(col_vals_filtered)]
            else:
                domain = [min(col_vals_filtered), max(col_vals_filtered)]

        # Rescale based on the given domain
        scaled_vals = _rescale_numeric(
            df=data_table,
            vals=col_vals,
            # Alternatively we could convert the domain to floats, but I dont mind ignoring this
            domain=domain,  # type: ignore
        )
    else:
        raise TypeError(
            f"Invalid column type provided ({col_name}). Please ensure that the column is numeric."
        )

    # Map scaled values back to original positions, using 0 for NAs
    scaled_vals_fixed = []
    for orig_val, scaled_val in zip(col_vals, scaled_vals):
        if is_na(data_table, orig_val):
            scaled_vals_fixed.append(0)

        elif is_na(data_table, scaled_val):
            # consider handling by leaving the original val, and having a third color/category.

            # If original value < domain[0], set to 0; if > domain[1], set to 1
            if orig_val < min(domain):
                warnings.warn(
                    f"Value {orig_val} in column '{col_name}' is less than the domain minimum {min(domain)}. Setting to {min(domain)}.",
                    category=UserWarning,
                )
                scaled_vals_fixed.append(0)

            else:
                warnings.warn(
                    f"Value {orig_val} in column '{col_name}' is greater than the domain maximum {max(domain)}. Setting to {max(domain)}.",
                    category=UserWarning,
                )
                scaled_vals_fixed.append(1)
        else:
            scaled_vals_fixed.append(scaled_val)

    return scaled_vals_fixed


def _format_numeric_text(value: float, num_decimals: int) -> str:
    if num_decimals == 0:
        return f"{value:.0f}"
    else:
        return f"{value:.{num_decimals}f}".rstrip("0").rstrip(".")
