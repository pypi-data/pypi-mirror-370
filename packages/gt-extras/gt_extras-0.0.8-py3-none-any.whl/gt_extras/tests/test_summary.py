from datetime import datetime, timezone

import numpy as np
import pandas as pd
import polars as pl
import pytest
from great_tables import GT

from gt_extras.summary import gt_plt_summary
from gt_extras.tests.conftest import assert_rendered_body


def test_gt_plt_summary_snap(snapshot):
    for DataFrame in [pd.DataFrame, pl.DataFrame]:
        df = DataFrame(
            {
                "numeric": [1.5, 2.2, 3.3, None, 5.1],
                "string": ["A", "B", "A", "C", None],
                "boolean": [True, False, True, False, False],
                "datetime": [
                    datetime(2024, 1, 1, tzinfo=timezone.utc),
                    datetime(2024, 1, 2, tzinfo=timezone.utc),
                    datetime(2024, 1, 3, tzinfo=timezone.utc),
                    None,
                    datetime(2024, 1, 5, tzinfo=timezone.utc),
                ],
            }
        )
        res = gt_plt_summary(df)
        assert_rendered_body(snapshot(name="pd_and_pl"), gt=res)


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_basic(DataFrame):
    df = DataFrame(
        {
            "numeric": [1, 2, 3, 4, 5],
            "string": ["A", "B", "C", "D", "E"],
            "boolean": [True, False, True, False, True],
        }
    )

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert isinstance(result, GT)
    assert "Summary Table" in html
    assert "5 rows x 3 cols" in html
    assert html.count("<svg") == 6

    assert "numeric" in html
    assert "string" in html
    assert "boolean" in html

    assert "opacity: 0;" in html
    assert ":hover" in html
    assert "transition:" in html

    assert ">Type</th>" in html
    assert ">Column</th>" in html
    assert ">Plot Overview</th>" in html
    assert ">Missing</th>" in html
    assert ">Mean</th>" in html
    assert ">Median</th>" in html
    assert ">SD</th>" in html

    assert "[1 to 3]" in html
    assert "[3 to 5]" in html

    assert html.count(">1 row</text>") == 5
    assert html.count(">2 rows</text>") == 2
    assert html.count(">3 rows</text>") == 2

    assert html.count('fill="#4e79a7"') == 5
    assert html.count('fill="#a65773"') == 2
    assert html.count('fill="#f18e2c"') == 2

    assert html.count(">0.0%</td>") == 3
    assert html.count(">3.00</td>") == 2
    assert html.count(">1.58</td>") == 1
    assert html.count(">0.60</td>") == 1


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_custom_title(DataFrame):
    df = DataFrame({"col": [1, 2, 3]})

    result = gt_plt_summary(df, title="Custom Title")
    html = result.as_raw_html()

    assert (
        '<td colspan="7" class="gt_heading gt_title gt_font_normal">Custom Title</td>'
        in html
    )
    assert (
        '<td colspan="7" class="gt_heading gt_subtitle gt_font_normal gt_bottom_border">3 rows x 1 cols</td>'
        in html
    )


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_numeric_column(DataFrame):
    df = DataFrame({"numeric": [1.1, 2.2, 3.3, 4.4, 5.5]})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "<title>signal</title>" in html
    assert '<td class="gt_row gt_right">3.30</td>' in html
    assert "<svg" in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_string_column(DataFrame):
    df = DataFrame({"string": ["A", "B", "A", "C", "A"]})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "<title>List</title>" in html
    assert "<svg" in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_boolean_column(DataFrame):
    df = DataFrame({"boolean": [True, False, True, True, False]})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "<title>Check</title>" in html
    assert "0.60" in html
    assert "<svg" in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_datetime_column(DataFrame):
    df = DataFrame(
        {
            "datetime": [
                datetime(2024, 1, 1, tzinfo=timezone.utc),
                datetime(2024, 1, 2, tzinfo=timezone.utc),
                datetime(2024, 1, 3, tzinfo=timezone.utc),
                datetime(2024, 1, 4, tzinfo=timezone.utc),
                datetime(2024, 1, 5, tzinfo=timezone.utc),
            ]
        }
    )

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "<svg" in html
    assert "<title>Clock</title>" in html
    assert "2024-01-01</text>" in html
    assert "2024-01-05</text>" in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_with_missing_values(DataFrame):
    df = DataFrame(
        {
            "numeric": [1, 2, None, 4, 5],
            "string": ["A", None, "C", "D", "E"],
            "boolean": [True, False, None, False, True],
        }
    )

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert html.count("20.0%") == 3


def test_gt_plt_summary_with_nan_values():
    df = pd.DataFrame(
        {
            "numeric": [1.0, 2.0, np.nan, 4.0, 5.0],
            "mixed": [1, np.nan, None, 4, 5],
        }
    )

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "20.0%" in html
    assert "40.0%" in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_empty_dataframe(DataFrame):
    df = DataFrame()

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "0 rows x 0 cols" in html
    assert "<svg" not in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_empty_columns(DataFrame):
    df = DataFrame({"col1": [], "col2": []})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "0 rows x 2 cols" in html
    assert html.count("100.0%") == 2


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_single_row(DataFrame):
    df = DataFrame(
        {
            "numeric": [42],
            "string": ["test"],
            "boolean": [True],
        }
    )

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "1 rows x 3 cols" in html
    assert html.count("42.00") == 2

    assert html.count(">1 row</text>") == 3
    assert html.count(">0 rows</text>") == 2

    assert '>"True"</text>' in html
    assert '>"False"</text>' not in html


def test_gt_plt_summary_all_missing_column():
    # Pandas version
    df_pd = pd.DataFrame(
        {
            "all_none": [None, None, None],
            "all_nan": [np.nan, np.nan, np.nan],
            "mixed": [np.nan, None, float("nan")],
        }
    )

    result = gt_plt_summary(df_pd)
    html = result.as_raw_html()

    assert "3 rows x 3 cols" in html
    assert html.count("100.0%") == 3

    # Polars version
    df_pl = pl.DataFrame(
        {
            "all_none": [None, None, None],
            "mixed": [None, None, None],
        }
    )

    result = gt_plt_summary(df_pl)
    html = result.as_raw_html()

    assert "3 rows x 2 cols" in html
    assert html.count("100.0%") == 2


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_single_value_numeric(DataFrame):
    df = DataFrame({"numeric": [5, 5, 5, 5, 5]})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "[3.5 to 4.5]" in html
    assert "[4.5 to 5.5]" in html
    assert "[5.5 to 6.5]" in html

    assert html.count(">0 rows</text>") == 2
    assert html.count(">5 rows</text>") == 1


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_boolean_all_true(DataFrame):
    df = DataFrame({"boolean": [True, True, True]})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "1.00" in html
    assert 'fill-opacity="1.0"' in html
    assert ">0.0%</td>" in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_boolean_all_false(DataFrame):
    df = DataFrame({"boolean": [False, False, False]})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "0.00" in html
    assert 'fill-opacity="0.2"' in html
    assert ">0.0%</td>" in html


def test_gt_plt_summary_boolean_all_empty():
    df = pl.DataFrame({"boolean": [None, None, None]}, schema={"boolean": bool})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "<div></div>" in html


def test_gt_plt_summary_datetime_single_date():
    df = pl.DataFrame(
        {"datetime": [datetime(2024, 1, 1, tzinfo=timezone.utc)]},
        schema={"datetime": datetime},
    )

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "[2023-12-30 to 2023-12-31]" in html
    assert "[2023-12-31 to 2024-01-01]" in html
    assert "[2024-01-01 to 2024-01-02]" in html

    assert html.count(">0 rows</text>") == 2
    assert html.count(">1 row</text>") == 1


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_datetime_two_dates(DataFrame):
    df = DataFrame(
        {
            "datetime": [
                datetime(2024, 1, 1, tzinfo=timezone.utc),
                datetime(2024, 1, 2, tzinfo=timezone.utc),
            ]
        }
    )

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert html.count(">1 row</text>") == 2

    assert "[2024-01-01 to 2024-01-01]" in html
    assert "[2024-01-01 to 2024-01-02]" in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_datetime_repeated_date(DataFrame):
    df = DataFrame(
        {
            "datetime": [
                datetime(2024, 1, 1, tzinfo=timezone.utc),
                datetime(2024, 1, 1, tzinfo=timezone.utc),
                datetime(2024, 1, 1, tzinfo=timezone.utc),
            ]
        }
    )

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "[2023-12-30 to 2023-12-31]" in html
    assert "[2023-12-31 to 2024-01-01]" in html
    assert "[2024-01-01 to 2024-01-02]" in html

    assert html.count(">0 rows</text>") == 2
    assert html.count(">3 rows</text>") == 1


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_numeric_two_values(DataFrame):
    df = DataFrame({"numeric": [1, 2]})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert html.count(">1 row</text>") == 2

    assert "[1 to 1.5]" in html
    assert "[1.5 to 2]" in html


def test_gt_plt_summary_unknown_dtype():
    df = pd.DataFrame({"other": [object(), object(), object()]})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert "<title>Question</title>" in html
    assert "other" in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_negative_numbers(DataFrame):
    df = DataFrame({"negative": [-5, -3, -1, 1, 3]})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert html.count("−1.00") == 2
    assert html.count("3.16") == 1

    assert "[-5 to -1]" in html
    assert "[-1 to 3]" in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_categorical_many_categories(DataFrame):
    categories = [f"Cat_{i}" for i in range(20)]
    df = DataFrame({"many_cats": categories})

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    assert html.count('x="21.8"') == 4
    assert html.count('x="155.8"') == 4

    for i in range(20):
        assert f'>"Cat_{i}"</text>' in html


@pytest.mark.parametrize("DataFrame", [pd.DataFrame, pl.DataFrame])
def test_gt_plt_summary_column_order_preserved(DataFrame):
    df = DataFrame(
        {
            "z_column": [1, 2, 3],
            "a_column": ["A", "B", "C"],
            "m_column": [True, False, True],
        }
    )

    result = gt_plt_summary(df)
    html = result.as_raw_html()

    z_pos = html.find("z_column")
    a_pos = html.find("a_column")
    m_pos = html.find("m_column")

    assert z_pos < a_pos < m_pos


# TODO: time
# def test_gt_plt_summary_datetime_with_time():
#     df = pd.DataFrame(
#         {
#             "datetime_with_time": [
#                 datetime(2024, 1, 1, 10, 30, 0, tzinfo = timezone.utc),
#                 datetime(2024, 1, 1, 14, 45, 30, tzinfo = timezone.utc),
#                 datetime(2024, 1, 2, 9, 15, 45, tzinfo = timezone.utc),
#             ]
#         }
#     )

#     result = gt_plt_summary(df)
#     html = result.as_raw_html()

#     assert "2024-01-01" in html
