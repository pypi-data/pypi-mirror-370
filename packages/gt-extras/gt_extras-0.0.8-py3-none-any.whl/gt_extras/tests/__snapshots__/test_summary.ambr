# serializer version: 1
# name: test_gt_plt_summary_snap[pd_and_pl]
  '''
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_left"><svg viewBox="0 0 576 512" preserveAspectRatio="none" aria-label="signal" role="img" class="fa" style="fill:#f18e2c;height:17.78px;width:20px;margin-left:auto;margin-right:0.2em;position:relative;vertical-align:-0.125em;overflow:visible;">  <title>signal</title>
    <path d="M544 0c17.7 0 32 14.3 32 32V480c0 17.7-14.3 32-32 32s-32-14.3-32-32V32c0-17.7 14.3-32 32-32zM416 96c17.7 0 32 14.3 32 32V480c0 17.7-14.3 32-32 32s-32-14.3-32-32V128c0-17.7 14.3-32 32-32zM320 224V480c0 17.7-14.3 32-32 32s-32-14.3-32-32V224c0-17.7 14.3-32 32-32s32 14.3 32 32zM160 288c17.7 0 32 14.3 32 32V480c0 17.7-14.3 32-32 32s-32-14.3-32-32V320c0-17.7 14.3-32 32-32zM64 416v64c0 17.7-14.3 32-32 32s-32-14.3-32-32V416c0-17.7 14.3-32 32-32s32 14.3 32 32z"></path></svg></td>
      <td style="font-weight: bold;" class="gt_row gt_left">numeric</td>
      <td class="gt_row gt_center"><svg xmlns="http://www.w3.org/2000/svg" width="180" height="40"><rect stroke="transparent" class="id0-hover-area" id="id0-hover-area-1" x="94.275" y="0" width="76.95" height="-0.5333333333333333" fill="transparent"/><rect class="id0-bar-rect" id="id0-bar-1" x="94.275" y="-0.5333333333333333" width="76.95" height="32.0" fill="#f18e2c"/><rect stroke="transparent" class="id0-hover-area" id="id0-hover-area-0" x="8.775" y="0" width="76.95" height="-0.5333333333333333" fill="transparent"/><rect class="id0-bar-rect" id="id0-bar-0" x="8.775" y="-0.5333333333333333" width="76.95" height="32.0" fill="#f18e2c"/><style>
      .id0-tooltip {
          opacity: 0;
          transition: opacity 0.2s;
          pointer-events: none;
      }
      .id0-bar-rect:hover, .id0-visual-bar:hover {
          stroke: white; stroke-width: 1.0666666666666667; fill-opacity: 0.6;
      }
      #id0-bar-0:hover ~ #id0-tooltip-0 { opacity: 1; }
  #id0-hover-area-0:hover ~ #id0-tooltip-0 { opacity: 1; }
  #id0-hover-area-0:hover ~ #id0-bar-0 { stroke: white; stroke-width: 1.0666666666666667; fill-opacity: 0.6; }
  #id0-bar-1:hover ~ #id0-tooltip-1 { opacity: 1; }
  #id0-hover-area-1:hover ~ #id0-tooltip-1 { opacity: 1; }
  #id0-hover-area-1:hover ~ #id0-bar-1 { stroke: white; stroke-width: 1.0666666666666667; fill-opacity: 0.6; }</style><line stroke="black" stroke-width="1.0666666666666667" x1="0" y1="32.0" x2="180" y2="32.0"/><line stroke="black" stroke-width="1.0666666666666667" x1="76.9375" y1="31.466666666666665" x2="76.9375" y2="-0.5333333333333333"/><text dominant-baseline="text-top" text-anchor="middle" font-size="8.0" x="47.25" y="40">1.5</text><text dominant-baseline="text-top" text-anchor="middle" font-size="8.0" x="132.75" y="40">5.1</text><g id="id0-tooltip-0" class="id0-tooltip"><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="47.25" y="2.0">2 rows</text><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="47.25" y="12.0">[1.5 to 3.3]</text></g><g id="id0-tooltip-1" class="id0-tooltip"><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="132.75" y="2.0">2 rows</text><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="132.75" y="12.0">[3.3 to 5.1]</text></g></svg></td>
      <td class="gt_row gt_right">20.0%</td>
      <td class="gt_row gt_right">3.02</td>
      <td class="gt_row gt_right">2.75</td>
      <td class="gt_row gt_right">1.57</td>
    </tr>
    <tr>
      <td class="gt_row gt_left gt_striped"><svg viewBox="0 0 512 512" preserveAspectRatio="none" aria-label="List" role="img" class="fa" style="fill:#4e79a7;height:20.0px;width:20px;margin-left:auto;margin-right:0.2em;position:relative;vertical-align:-0.125em;overflow:visible;">  <title>List</title>
    <path d="M40 48C26.7 48 16 58.7 16 72v48c0 13.3 10.7 24 24 24H88c13.3 0 24-10.7 24-24V72c0-13.3-10.7-24-24-24H40zM192 64c-17.7 0-32 14.3-32 32s14.3 32 32 32H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H192zm0 160c-17.7 0-32 14.3-32 32s14.3 32 32 32H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H192zm0 160c-17.7 0-32 14.3-32 32s14.3 32 32 32H480c17.7 0 32-14.3 32-32s-14.3-32-32-32H192zM16 232v48c0 13.3 10.7 24 24 24H88c13.3 0 24-10.7 24-24V232c0-13.3-10.7-24-24-24H40c-13.3 0-24 10.7-24 24zM40 368c-13.3 0-24 10.7-24 24v48c0 13.3 10.7 24 24 24H88c13.3 0 24-10.7 24-24V392c0-13.3-10.7-24-24-24H40z"></path></svg></td>
      <td style="font-weight: bold;" class="gt_row gt_left gt_striped">string</td>
      <td class="gt_row gt_center gt_striped"><svg xmlns="http://www.w3.org/2000/svg" width="180" height="40"><style>
      .id1-category-tooltip {
          opacity: 0;
          transition: opacity 0.2s;
          pointer-events: none;
      }
      .id1-bar-rect:hover, .id1-visual-bar:hover {
          opacity: 0.4;
      }
      #id1-bar-0:hover ~ #id1-tooltip-0 { opacity: 1; }
  #id1-bar-1:hover ~ #id1-tooltip-1 { opacity: 1; }
  #id1-bar-2:hover ~ #id1-tooltip-2 { opacity: 1; }</style><rect stroke="transparent" class="id1-visual-bar" id="id1-bar-2" x="132.75" y="4.0" width="42.75" height="32.0" fill-opacity="0.19999999999999996" fill="#4e79a7"/><rect stroke="transparent" class="id1-visual-bar" id="id1-bar-1" x="90.0" y="4.0" width="42.75" height="32.0" fill-opacity="0.6" fill="#4e79a7"/><rect stroke="transparent" class="id1-visual-bar" id="id1-bar-0" x="4.5" y="4.0" width="85.5" height="32.0" fill-opacity="1.0" fill="#4e79a7"/><g id="id1-tooltip-0" class="id1-category-tooltip"><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="47.25" y="10.0">2 rows</text><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="47.25" y="20.0">"A"</text></g><g id="id1-tooltip-1" class="id1-category-tooltip"><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="111.375" y="10.0">1 row</text><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="111.375" y="20.0">"B"</text></g><g id="id1-tooltip-2" class="id1-category-tooltip"><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="154.125" y="10.0">1 row</text><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="154.125" y="20.0">"C"</text></g></svg></td>
      <td class="gt_row gt_right gt_striped">20.0%</td>
      <td class="gt_row gt_right gt_striped">&mdash;</td>
      <td class="gt_row gt_right gt_striped">&mdash;</td>
      <td class="gt_row gt_right gt_striped">&mdash;</td>
    </tr>
    <tr>
      <td class="gt_row gt_left"><svg viewBox="0 0 512 512" preserveAspectRatio="none" aria-label="Check" role="img" class="fa" style="fill:#a65773;height:20.0px;width:20px;margin-left:auto;margin-right:0.2em;position:relative;vertical-align:-0.125em;overflow:visible;">  <title>Check</title>
    <path d="M470.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L192 338.7 425.4 105.4c12.5-12.5 32.8-12.5 45.3 0z"></path></svg></td>
      <td style="font-weight: bold;" class="gt_row gt_left">boolean</td>
      <td class="gt_row gt_center"><svg xmlns="http://www.w3.org/2000/svg" width="180" height="40"><style>
      .id2-category-tooltip {
          opacity: 0;
          transition: opacity 0.2s;
          pointer-events: none;
      }
      .id2-bar-rect:hover, .id2-visual-bar:hover {
          opacity: 0.4;
      }
      #id2-bar-0:hover ~ #id2-tooltip-0 { opacity: 1; }
  #id2-bar-1:hover ~ #id2-tooltip-1 { opacity: 1; }</style><rect stroke="transparent" class="id2-visual-bar" id="id2-bar-1" x="72.9" y="4.0" width="102.6" height="32.0" fill-opacity="0.2" fill="#a65773"/><rect stroke="transparent" class="id2-visual-bar" id="id2-bar-0" x="4.5" y="4.0" width="68.4" height="32.0" fill-opacity="1.0" fill="#a65773"/><g id="id2-tooltip-0" class="id2-category-tooltip"><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="38.7" y="10.0">2 rows</text><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="38.7" y="20.0">"True"</text></g><g id="id2-tooltip-1" class="id2-category-tooltip"><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="124.2" y="10.0">3 rows</text><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="124.2" y="20.0">"False"</text></g></svg></td>
      <td class="gt_row gt_right">0.0%</td>
      <td class="gt_row gt_right">0.40</td>
      <td class="gt_row gt_right">&mdash;</td>
      <td class="gt_row gt_right">&mdash;</td>
    </tr>
    <tr>
      <td class="gt_row gt_left gt_striped"><svg viewBox="0 0 512 512" preserveAspectRatio="none" aria-label="Clock" role="img" class="fa" style="fill:#73a657;height:20.0px;width:20px;margin-left:auto;margin-right:0.2em;position:relative;vertical-align:-0.125em;overflow:visible;">  <title>Clock</title>
    <path d="M256 512C114.6 512 0 397.4 0 256S114.6 0 256 0S512 114.6 512 256s-114.6 256-256 256zM232 120V256c0 8 4 15.5 10.7 20l96 64c11 7.4 25.9 4.4 33.3-6.7s4.4-25.9-6.7-33.3L280 243.2V120c0-13.3-10.7-24-24-24s-24 10.7-24 24z"></path></svg></td>
      <td style="font-weight: bold;" class="gt_row gt_left gt_striped">datetime</td>
      <td class="gt_row gt_center gt_striped"><svg xmlns="http://www.w3.org/2000/svg" width="180" height="40"><rect stroke="transparent" class="id3-hover-area" id="id3-hover-area-1" x="94.275" y="0" width="76.95" height="-0.5333333333333333" fill="transparent"/><rect class="id3-bar-rect" id="id3-bar-1" x="94.275" y="-0.5333333333333333" width="76.95" height="32.0" fill="#73a657"/><rect stroke="transparent" class="id3-hover-area" id="id3-hover-area-0" x="8.775" y="0" width="76.95" height="-0.5333333333333333" fill="transparent"/><rect class="id3-bar-rect" id="id3-bar-0" x="8.775" y="-0.5333333333333333" width="76.95" height="32.0" fill="#73a657"/><style>
      .id3-tooltip {
          opacity: 0;
          transition: opacity 0.2s;
          pointer-events: none;
      }
      .id3-bar-rect:hover, .id3-visual-bar:hover {
          stroke: white; stroke-width: 1.0666666666666667; fill-opacity: 0.6;
      }
      #id3-bar-0:hover ~ #id3-tooltip-0 { opacity: 1; }
  #id3-hover-area-0:hover ~ #id3-tooltip-0 { opacity: 1; }
  #id3-hover-area-0:hover ~ #id3-bar-0 { stroke: white; stroke-width: 1.0666666666666667; fill-opacity: 0.6; }
  #id3-bar-1:hover ~ #id3-tooltip-1 { opacity: 1; }
  #id3-hover-area-1:hover ~ #id3-tooltip-1 { opacity: 1; }
  #id3-hover-area-1:hover ~ #id3-bar-1 { stroke: white; stroke-width: 1.0666666666666667; fill-opacity: 0.6; }</style><line stroke="black" stroke-width="1.0666666666666667" x1="0" y1="32.0" x2="180" y2="32.0"/><line stroke="black" stroke-width="1.0666666666666667" x1="79.3125" y1="31.466666666666665" x2="79.3125" y2="-0.5333333333333333"/><text dominant-baseline="text-top" text-anchor="middle" font-size="8.0" x="47.25" y="40">2024-01-01</text><text dominant-baseline="text-top" text-anchor="middle" font-size="8.0" x="132.75" y="40">2024-01-05</text><g id="id3-tooltip-0" class="id3-tooltip"><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="52.0" y="2.0">2 rows</text><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="52.0" y="12.0">[2024-01-01 to 2024-01-03]</text></g><g id="id3-tooltip-1" class="id3-tooltip"><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="128.0" y="2.0">2 rows</text><text dominant-baseline="hanging" text-anchor="middle" font-size="8.0" font-weight="bold" fill="black" x="128.0" y="12.0">[2024-01-03 to 2024-01-05]</text></g></svg></td>
      <td class="gt_row gt_right gt_striped">20.0%</td>
      <td class="gt_row gt_right gt_striped">&mdash;</td>
      <td class="gt_row gt_right gt_striped">&mdash;</td>
      <td class="gt_row gt_right gt_striped">&mdash;</td>
    </tr>
  </tbody>
  '''
# ---
